dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/cache/impl/TaskInstanceCacheManagerImpl.java

```diff
@@ -57,7 +86,7 @@
     @Override
     public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
         TaskInstance taskInstance = taskInstanceCache.get(taskInstanceId);
```

Comment:
It is better to use `computeIfAbsent` here.

Reply:
I put database access out of `computeIfAbsent`'s `Function` class `apply` method, is this ok?

Reply:
I put database access out of `computeIfAbsent`'s `Function` class `apply` method, is this ok?

Reply:
@blackberrier I think you can change the entire method to :), 
```
return taskInstanceCache.computeIfAbsent(taskInstanceId, k ->processService.findTaskInstanceById(k));
```

Reply:
@blackberrier You can change the entire method to 
```
return taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
```

Reply:
@ruanwenjun 
I think the database may waste time or something, and thus block other thread. So I take the  `processService.findTaskInstanceById(k)` out

Reply:
I know your meaning. I have checked the comment of the `computeIfAbsent`, it says as follows. So I wonder if the database access operation should take out.

![1622725677097](https://user-images.githubusercontent.com/6926304/120649848-c8ac7b80-c4af-11eb-9513-7805180fb339.jpg)

Reply:
@blackberrier 
If there are two threads execute `getByTaskInstanceId` at the same time, we just want to query from database once, right? The block will only happen on the two thread update on the same key, and the key is not exist, this is expected. 

And you should realize that the below way is not an atomic operation, in some situation, it might cause problem.
```java
TaskInstance taskInstance = taskInstanceCache.get(taskInstanceId);
if (taskInstance == null){
    taskInstance = processService.findTaskInstanceById(taskInstanceId);
    taskInstanceCache.put(taskInstanceId,taskInstance);
}
return taskInstance;
```

Reply:
@ruanwenjun 
I think the block may happen on two thread on two key, and the two key share the same hashcode, which means they are on the same Linkedlist or Red-black tree, and in `computeIfAbsent` synchronized lock on the head node. 
If we write like this,
```
taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
```
and if the first key access database and wasting time , then the second key have to wait.

If we take database access out, like
```
taskInstance = processService.findTaskInstanceById(taskInstanceId);
TaskInstance finalTaskInstance = taskInstance;
taskInstanceCache.computeIfAbsent(taskInstanceId, k -> finalTaskInstance);
```
the second key need not wait while the first key is accessing database.

I wonder if my thought is reasonable.

Reply:
@blackberrier 

> the block may happen on two thread on two key, and the two key share the same hashcode

Yes, but we can avoid the 1 step.
```
1. taskInstance = processService.findTaskInstanceById(taskInstanceId);
2. TaskInstance finalTaskInstance = taskInstance;
3. taskInstanceCache.computeIfAbsent(taskInstanceId, k -> finalTaskInstance);
```

Reply:
@blackberrier I mean you can use below code
```
public TaskInstance getByTaskInstanceId(Integer taskInstanceId) {
return taskInstanceCache.computeIfAbsent(taskInstanceId, k -> processService.findTaskInstanceById(k));
}
```
The cache is used to reduce database pressure, in your plan, if there are multiple threads query the same taskInstance at the same time, they will all query from database.
Your consideration is that when conflicts occur, multiple thread query will block. I think if conflicts occur, if we don't want to be blocked, we should find a better hash method to reduce conflicts. 

In additional, we can use one line of code to achieve, why use more complex code?

Reply:
@ruanwenjun 
ok, I have commit latest code